LuaUI/Widgets/cmd_hold_ctrl_to_place_and_morph.lua

```diff
@@ -0,0 +1,126 @@
+---@alias UnitId integer
+
+---@param names UnitInternalName[]
+---@return table<UnitDefId, boolean>
+local function CreateUnitDefIdSet(names)
+	local result = {}
+	for _, name in ipairs(names) do
+		result[UnitDefNames[name].id] = true
+	end
+	return result
+end
+
+local abs = math.abs
+local spGetUnitPosition = Spring.GetUnitPosition
+local spGiveOrderToUnit = Spring.GiveOrderToUnit
+
+local myTeamID
+
+local morphableUnitDefIds = CreateUnitDefIdSet({'staticjammer', 'staticshield', 'staticradar', 'energygeo'})
+
+---@type table<UnitId,{x:number,z:number}[]>
+local buildingsToMorphByBuilder = {}
+
+function widget:Initialize()
+	myTeamID = Spring.GetMyTeamID()
+end
+
+function widget:UnitCommand(unitID, unitDefID, unitTeam, cmdId, cmdParams, cmdOpts, cmdTag)
+	if cmdId > 0 
+		or unitTeam ~= myTeamID 
+		or not morphableUnitDefIds[-cmdId] 
+		or not cmdOpts
+	then 
+		return 
+	end
+
+	local buildingsToMorph = buildingsToMorphByBuilder[unitID]
+	local point = { x = cmdParams[1], z = cmdParams[3] }
+
+	if cmdOpts.ctrl then
+		if not buildingsToMorph then
+			buildingsToMorph = {}
+			buildingsToMorphByBuilder[unitID] = buildingsToMorph
+		end
+		table.insert(buildingsToMorph, point)
+
+		-- Spring.Echo('Building at '..show(point).. ' marked as to be morphed when placed by builder '..unitID)
+
+	elseif buildingsToMorph then
+		for i, point2 in pairs(buildingsToMorph) do
+			if (point2.x) then
+				if (point2.x == point.x) and (point2.z == point.z) then
+					buildingsToMorph[i] = nil
+				end
+			end
+		end
+	end
+end
+
+function widget:UnitCreated(unitID, unitDefID, unitTeam, builderID)
+	if not morphableUnitDefIds[unitDefID] 
+		or unitTeam ~= myTeamID 
+		or not builderID then
+		return
+	end
+
+	local buildingsToMorph = buildingsToMorphByBuilder[builderID]
+
+	if not buildingsToMorph then 
+		-- Spring.Echo('No buildings to morph!')
+		return
+	end
+
+
+	local ux, uy, uz  = spGetUnitPosition(unitID)
+	-- Spring.Echo('UnitCreated(unitDef = '..UnitDefs[unitDefID].name..', builderID = '..(builderID or 'nil')..', x = '..ux..', z = '..uz..')')
+
+	for i, point2 in pairs(buildingsToMorph) do
+		if (point2.x) then
+
+			-- Note: unit_building_starter.lua, which does something similar, uses a location tolerance of 16
+			-- here. But this appears to be unnecessary for the morphable buildings considered here(?)
+			-- (in fact, it workers with exact equality, but being defensive)
+
+			if abs(point2.x - ux) < 1e-3 and abs(point2.z - uz) < 1e-3 then
+				local cmdId = Spring.Utilities.CMD.MORPH
+				-- Spring.Echo('MORPHING!!!!!!! '..cmdId)
+				spGiveOrderToUnit(unitID, cmdId, {}, 0)
+				return
+			end
+		end
+	end
+
+	-- Spring.Echo('This building is not to be morphed!')
+end
+
+function widget:UnitDestroyed(unitID, unitDefID, unitTeam)
+	if buildingsToMorphByBuilder[unitID] then
+		-- Spring.Echo('Clearing buildingsToMorph for builder '..unitID..' because it was destroyed.')
+		buildingsToMorphByBuilder[unitID] = nil
+	end
```

Comment:
Avoiding the needless assignment is as slow as the assignment itself, so offers no benefit and doubles the work in case the value wasn't already `nil`.
```suggestion
	buildingsToMorphByBuilder[unitID] = nil
```